### ISSUE TITLE
Setting the Viewport's scaling_3d_scale to any value except 1.0 while the WorldEnvironment's background_mode is set to BG_CANVAS breaks Z sorting in Compatibility renderer.

### ISSUE THREAD

**Issue by atomius0** [godotengine/godot#97039]
### Tested versions

- Reproducible in: v4.3.stable.official [77dcf97d8], v4.4.dev2.official [97ef3c837]

### System information

Godot v4.3.stable - Fedora Linux 37 (KDE Plasma) - X11 - GLES3 (Compatibility) - AMD Radeon RX 6650 XT (navi23, LLVM 15.0.7, DRM 3.49, 6.2.8-200.fc37.x86_64) - AMD Ryzen 5 5600X 6-Core Processor (12 Threads)

### Issue description

In a 3D scene, the Z sorting(?) breaks when setting the Viewport's "scaling_3d_scale" (or "rendering/scaling_3d/scale" in the project settings) to any value except 1.0 while the WorldEnvironment's background_mode is set to "Canvas" (Environment.BG_CANVAS).

Setting the scaling_3d_scale back to 1.0, or changing the background_mode to "Clear Color" (Environment.BG_CLEAR_COLOR) makes the Z sorting go back to normal.

This issue only occurs with the Compatibility renderer. (I tested it on Linux, Windows and in Web exports)
It does not occur when using the Forward+ or Mobile renderers.

The same issue occurs with these configurations, too (the System information field only allows one line, so I'll paste them here):
- Godot v4.4.dev2 - Fedora Linux 37 (KDE Plasma) - X11 - OpenGL 3 (Compatibility) - AMD Radeon RX 6650 XT (navi23, LLVM 15.0.7, DRM 3.49, 6.2.8-200.fc37.x86_64) - AMD Ryzen 5 5600X 6-Core Processor (12 Threads)

- Godot v4.3.stable - Windows 6.1.7601 - GLES3 (Compatibility) - GeForce GTX 750 Ti - AMD FX(tm)-6350 Six-Core Processor (6 Threads)

### Steps to reproduce

- run the minimal reproduction project: everything looks normal.
![Screenshot_20240915_144121](https://github.com/user-attachments/assets/215b5182-0baf-4659-82ac-2fc9f10cf9bf)

- drag the "scaling_3d_scale" slider to any value other than 1.0: Z sorting breaks.
![Screenshot_20240915_144131](https://github.com/user-attachments/assets/fb1253e8-b2e9-4c4f-9c2e-e2e7764a23b9)

- change "background_mode" to "Clear Color": it looks fine again, no matter what scale it is set to.
![Screenshot_20240915_144135](https://github.com/user-attachments/assets/e984475a-615c-449e-9525-6baeeb4803dd)

### Minimal reproduction project (MRP)

[env_bg_canvas_z_bug.zip](https://github.com/user-attachments/files/17005851/env_bg_canvas_z_bug.zip)

**Comment by matheusmdx** [comment]
Bisecting points to #87360 as the culprit, @BastiaanOlij

![image](https://github.com/user-attachments/assets/793834b1-be73-4555-9664-916a608f57a7)

**Comment by BastiaanOlij** [comment]
GRMBL dumb global state in OpenGL.....

One thing we fixed in that PR is that we copy the canvas into the background which was missing. But I made a booboo:

https://github.com/BastiaanOlij/godot/blob/aa260e5f3d4ce0c5551e84ef0b098ddf9e78fd0d/drivers/gles3/rasterizer_scene_gles3.cpp#L2612

Notice that we turn the depth test off before we copy the canvas into our background. 

We never turn it back on again...

oops.
